src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs

```diff
@@ -349,12 +350,39 @@ public void ActivityStart(string SourceName, string ActivityName, IEnumerable<Ke
         public void ActivityStop(string SourceName, string ActivityName, IEnumerable<KeyValuePair<string, string?>> Arguments) =>
             WriteEvent(12, SourceName, ActivityName, Arguments);
 
+        /// <summary>
+        /// Used to send version information.
+        /// </summary>
+        [Event(13, Keywords = Keywords.Messages)]
+        public void Version(int Major, int Minor, int Patch)
+        {
+            WriteEvent(13, Major, Minor, Patch);
+        }
+
         /// <summary>
         /// Called when the EventSource gets a command from a EventListener or ETW.
         /// </summary>
         [NonEvent]
         protected override void OnEventCommand(EventCommandEventArgs command)
         {
+            if (command.Command == EventCommand.Enable
+                && IsEnabled(EventLevel.Informational, Keywords.Messages))
+            {
+                var assembly = typeof(Activity).Assembly;
+
+                var version = assembly.GetCustomAttribute<AssemblyFileVersionAttribute>()?.Version
```

Comment:
@vitek-karas @ericstj @MichalStrehovsky do you see any issue using `Assembly.GetCustomAttribute` for AOT builds? 

CC @jkotas @ericstj

Reply:
It is very expensive way to compute a constant. Can it be embedded as a constant instead?

Reply:
We'll need to maintain the constant with every release of the library. Getting something automated will be much better I guess.

Reply:
RIght, I did not mean to hardcode the constant. I meant to propagate it from the build.

Reply:
@ericstj pointed out corelib does similar thing through the source gen https://github.com/dotnet/runtime/blob/18eedbe65fe5b1946f8c6df0a70357b7ed01a884/src/libraries/System.Private.CoreLib/gen/ProductVersionInfoGenerator.cs#L25. But I am not sure if it is worth it to do so. I am hoping for a simpler way if possible. by the way, if we go with `GetCustomAttribute`, it will be executed once when establishing the event source session for the library. My only concern using `GetCustomAttribute` if it can cause any issue for AOT which I am asking about.

Reply:
Here's a few other samples where folks use an MSBuild task to generate a file - https://github.com/search?q=org%3Adotnet+GenerateFileFromTemplate++language%3AXML&type=code&l=XML  Might be a slightly smaller change to do that if you wanted.

Reply:
Nice!

@CodeBlanch do you want to try it?

Reply:
Updated!